clustermesh-apiserver/etcdinit/root.go

```diff
@@ -0,0 +1,170 @@
+
+	// When the clustermesh-apiserver is launched, we create a new etcd. We don't support persistence of any kind
+	// here. So we just delete the contents of the data directory. It should be empty, but if the initialization failed
+	// we might re-run this as an init container.
+	log.Info("Deleting contents of data directory")
+	// We don't use os.RemoveAll on the etcdDataDirectory because we don't want to remove the directory itself, just
+	// everything inside of it. In most cases that directory will be a mount anyway, and so removing it might cause
+	// problems.
+	dir, err := os.ReadDir(etcdDataDir)
+	if err != nil {
+		log.WithError(err).Error("Failed to read from the etcd data directory while attempting to delete existing files")
+		return err
+	}
+	for _, d := range dir {
+		log.WithField("path", d.Name()).Debug("Removing file/directory in data dir")
+		err = os.RemoveAll(path.Join(etcdDataDir, d.Name()))
+		if err != nil {
+			log.WithField("path", d.Name()).WithError(err).Error("Failed to remove pre-existing file/directory in etcd data directory")
+			return err
+		}
+	}
+
+	// We use "localhost" so it works in both the IPv4 and IPv6 cases.
+	loopbackEndpoint := "http://localhost:2379"
+	log.WithField("etcdListenClientUrl", loopbackEndpoint).Info("Starting localhost-only etcd process")
+	// Specify the full path to the etcd binary to avoid any PATH search binary replacement nonsense
+	etcdCmd := exec.CommandContext(ctx, etcdBinaryLocation,
+		fmt.Sprintf("--data-dir=%s", etcdDataDir),
+		fmt.Sprintf("--name=%s", etcdClusterName),
+		fmt.Sprintf("--listen-client-urls=%s", loopbackEndpoint),
+		fmt.Sprintf("--advertise-client-urls=%s", loopbackEndpoint),
+		fmt.Sprintf("--initial-cluster-token=%s", etcdInitialClusterToken),
+		"--initial-cluster-state=new")
+	log.WithField("etcdBinary", etcdBinaryLocation).
+		WithField("etcdFlags", etcdCmd.Args).
+		Debug("Executing etcd")
+
+	// Exec the etcd binary, which ultimately calls fork(2) under the hood. We're attached to STDIN and STDOUT for the
+	// new process. We don't wait on its completion, because it'll never complete of course.
+	err = etcdCmd.Start()
+	if err != nil {
+		log.WithError(err).Error("Failed to launch etcd process")
+		return err
+	}
+	log = log.WithField("etcdPID", etcdCmd.Process.Pid)
+	log.Info("Local etcd server process started")
+
+	// Defer etcd process cleanup
+	defer func() {
+		log.Debug("Cleaning up etcd process")
+		// Send the process a SIGTERM. SIGTERM is the "gentle" shutdown signal, and etcd should close down it's resources
+		// cleanly and then exit.
+		log.Info("Sending SIGTERM signal to etcd process")
+		err = etcdCmd.Process.Signal(syscall.SIGTERM)
+		if err != nil {
+			log.WithError(err).Error("Failed to send SIGTERM signal to etcd process")
+		}
+
+		// Wait for the etcd process to finish, and cleanup resources.
+		log.Info("Waiting for etcd process to exit")
+		err = etcdCmd.Wait()
+		if err != nil {
```

Comment:
I'd explicitly check that the context didn't expire, and return an error in that case. Just to prevent missing the case in which the Wait call above got stuck and completed only because the context terminated (killing the etcd process).